src/analysis_and_optimization/Mem_pattern.ml

```diff
@@ -0,0 +1,597 @@
+ * functions and Assignments.
+ * @param in_loop A boolean to specify the logic of indexing expressions. See
+ *  `query_initial_demotable_expr` for an explanation of the logic.
+ *)
+let rec query_initial_demotable_stmt (in_loop : bool)
+    (Stmt.Fixed.({pattern; _}) :
+      (Expr.Typed.Meta.t, Stmt.Located.Meta.t) Stmt.Fixed.t) :
+    string Set.Poly.t =
+  let query_expr = query_initial_demotable_expr in_loop in
+  match pattern with
+  | Stmt.Fixed.Pattern.Assignment
+      ( ((name : string), (ut : UnsizedType.t), lhs)
+      , (Expr.Fixed.({meta= Expr.Typed.Meta.({type_; adlevel; _}); _}) as rhs)
+      ) ->
+      let check_lhs =
+        let lhs_list =
+          Set.Poly.union_list
+            (List.map
+               ~f:
+                 (Index.apply ~default:Set.Poly.empty ~merge:Set.Poly.union
+                    query_expr)
+               lhs)
+        in
+        match is_uni_eigen_loop_indexing in_loop ut lhs with
+        | true -> Set.Poly.add lhs_list name
+        | false -> lhs_list
+      in
+      let check_rhs = query_expr rhs in
+      let both_sides = Set.Poly.union check_lhs check_rhs in
+      let check_bad_assign =
+        match (UnsizedType.contains_eigen_type type_, adlevel) with
+        | true, UnsizedType.AutoDiffable ->
+            query_ad_real_data_matrix_assign rhs
+            || not (query_soa_supported_assign_expr rhs)
+        | _ -> false
+      in
+      if Set.Poly.length check_rhs > 0 || check_bad_assign then
+        Set.Poly.add both_sides name
+      else both_sides
+  | NRFunApp (kind, exprs) -> query_initial_demotable_funs in_loop kind exprs
+  | IfElse (predicate, lhs, rhs) ->
+      let query_rhs =
+        Option.value_map
+          ~f:(query_initial_demotable_stmt in_loop)
+          ~default:Set.Poly.empty rhs
+      in
+      Set.Poly.union_list
+        [ query_expr predicate
+        ; query_initial_demotable_stmt in_loop lhs
+        ; query_rhs ]
+  | Return optional_expr ->
+      Option.value_map ~f:query_expr ~default:Set.Poly.empty optional_expr
+  | SList lst | Profile (_, lst) | Block lst ->
+      Set.Poly.union_list
+        (List.map ~f:(query_initial_demotable_stmt in_loop) lst)
+  | TargetPE expr -> query_expr expr
+  | For {lower; upper; body; _} ->
+      Set.Poly.union
+        (Set.Poly.union (query_expr lower) (query_expr upper))
+        (query_initial_demotable_stmt true body)
+  | While (predicate, body) ->
+      Set.Poly.union (query_expr predicate)
+        (query_initial_demotable_stmt true body)
+  | Skip | Break | Continue | Decl _ -> Set.Poly.empty
+
+(* Look through a statement to see whether it needs modified from
+ * SoA to AoS. Returns the set of object names that need demoted
+ * in a statement, if any.
+ * @param aos_exits A set of variables that can be demoted.
+ * @param pattern The Stmt pattern to query.
+ *)
+let query_demotable_stmt (aos_exits : string Set.Poly.t)
```

Comment:
This is a really good point. Let me go back and check on this. I think this is going to require something a bit more clever

Reply:
So I _think_ this is fine because of how `SoA` only effects matrices. So for `size(A)` it does return an `int`, but for the purpose of deduction of what is allowable to be `SoA` that function returning back a non-matrix type is okay. Another example is like `array[] exp_quad_cov(vector, vector)`. So if we saw something like 

```stan
parameters {
  vector[N] Y;
  vector[N] Z;
}
transformed parameters {
  vector[N] X = to_vector(cov_exp_quad(Y, Z));
}
```

Here since `cov_exp_quad(Y, Z)` returns an array of reals we already know it's too expensive to convert that array of reals (structs) to SoA.

Reply:
@SteveBronder I'm still wondering about this, could you double check these test cases? Suppose `b` is SoA and `c` and `d` are AoS:
* `matrix a = size(b) > 1 ? c : d;`
* `matrix a = size(b) * c;`

Do we successfully demote `a` in those cases?

Reply:
I just added these as a test and they look good!

https://github.com/stan-dev/stanc3/pull/955/files#diff-58981cfa9e197e07f3bae4252ea70ce886c138fe9b479bcbfd185fbd9f004df2

Reply:
Is there a foolproof way of telling what's AoS vs. SoA? Looking at the CPP for that test [here](https://github.com/stan-dev/stanc3/blob/98c62e46777a9ca284f04e374e2ca9737b324215/test/integration/good/compiler-optimizations/mem_patterns/cpp.expected#L2581) and it looks like `b_soa` might be AoS, which would affect the results. Also appears to be AoS in an MIR dump:
```
 (((pattern
     (Decl (decl_adtype AutoDiffable) (decl_id b_soa)
      (decl_type
       (Sized
        (SMatrix AoS
         ((pattern (Lit Int 10))
          (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly))))
         ((pattern (Lit Int 10))
          (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly)))))))
      (initialize true)))
    (meta <opaque>))
```

Reply:
For that line you linked I'm seeing 

```
stan::conditional_var_value_t<local_scalar_t__, 
        Eigen::Matrix<local_scalar_t__, -1, -1>> b_soa
```

`stan::conditional_var_value_t` is a type trait that, if `local_scalar_t__` is `var` gives back a `var<Eigen::Matrix<local_scalar_t__, -1, -1>>` if `local_scalar_t__` is a `var` and otherwise gives back `Eigen::Matrix<local_scalar_t__, -1, -1>`. This is to handle `double` and `fvar<T>` as `local_scalar_t__`. Though what I'd like to do in the future is just have the tools around to write the `var` specialization of the `log_prob` function so we don't have to do that. 

The other one is a bug when I was printing out the transformed MIR (thanks!). I needed to print out the optimized transformed MIR. Now looks like how we expect!

```
((pattern
     (Decl (decl_adtype AutoDiffable) (decl_id b_soa)
      (decl_type
       (Sized
        (SMatrix SoA
         ((pattern (Lit Int 10))
          (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly))))
         ((pattern (Lit Int 10))
          (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly)))))))
      (initialize true)))
    (meta <opaque>))
```

https://github.com/stan-dev/stanc3/blob/f50fb6862b32342a2c4be50ca01aff48fc90b5b6/test/integration/good/compiler-optimizations/mem_patterns/transformed_mir.expected#L5887

Reply:
If I comment out the `cond_aos` line, I'm getting no demotions at all, can you replicate that?

I think the `cond_aos` line is getting added to the initial AoS set because it's setting [`is_all_rhs_aos`](https://github.com/stan-dev/stanc3/blob/98c62e46777a9ca284f04e374e2ca9737b324215/src/analysis_and_optimization/Mem_pattern.ml#L344) to true, which doesn't seem right, I think the `~set` and `~subset` are switched there. Then `cond_aos` being `AoS` demotes everything else

Reply:
Noting it here so I don't forget, it's worth checking `matrix a = transpose(c)` as well, since I'm not sure how `row(b) * c` demoted

Reply:
Pinging just in case the reply didn't, don't mean to rush you @SteveBronder

Reply:
Hey! Yes so this took me down a bit of a rabbit hole today. 

> I think the cond_aos line is getting added to the initial AoS set because it's setting is_all_rhs_aos to true, which doesn't seem right, I think the ~set and ~subset are switched there. Then cond_aos being AoS demotes everything else

Yes so I think those do need to be flipped.

I think for this line 

```stan
matrix[10, 10] int_aos_mul_aos = rows(row_soa) * empty_user_func_aos;
```

https://github.com/stan-dev/stanc3/pull/955/files#diff-3887bc12dee026edee6c3df0cde04b23cf6dd20949a84e78d04000fad60da677R27

it also didn't recognize in the initial variable sweep that `empty_user_func_aos` was AoS and so that lhs can also be demoted before even going to the monotone framework. So what I ended up doing was making the initial demotion search into a fold that keeps track of previously demoted values so then at this spot 

I added the stuff you posted above as a test [here](https://github.com/stan-dev/stanc3/pull/955/files#diff-3887bc12dee026edee6c3df0cde04b23cf6dd20949a84e78d04000fad60da677) and it's associated MIR is [here](https://github.com/stan-dev/stanc3/blob/e241be9c0ec061ca8eaa22e138025d02a8a4d4b7/test/integration/good/compiler-optimizations/mem_patterns/transformed_mir.expected#L11842). Just and fyi I try to label everything in those tests as `*_aos` or `*_soa` so if you ctrl+f for `SMatrix AoS` or `SMatrix SoA` the name in the decl should match up.

Reply:
@SteveBronder Gotcha. Thanks for adding the optimized MIR that makes it a lot easier. So it looks like we currently have issues with `empty_user_func_aos`, `int_aos_mul_aos` and `mul_two_aos`.

Any idea what's up with `empty_user_func_aos`? TBH I don't think I understand the fold fix.

It does look to me like the issue with `int_aos_mul_aos` and `mul_two_aos` is that `row_soa` is included as an SoA subexpression even though it's being used to produce a non-AD/eigen type. What if in `Mem_pattern.matrix_set` we don't recurse on non-AD/eigen expressions? Would that do it?

Reply:
> Gotcha. Thanks for adding the optimized MIR that makes it a lot easier. So it looks like we currently have issues with empty_user_func_aos, int_aos_mul_aos and mul_two_aos.

> Any idea what's up with empty_user_func_aos? TBH I don't think I understand the fold fix.

Yes sorry I missed those I think I have it fixed now. It was that I needed to move `is_not_supported_func` out of `is_all_rhs_aos` and use a `||` instead of a `&&`. I ctrl+f'd for `_aos` and those all seem to be good now. 

> What if in Mem_pattern.matrix_set we don't recurse on non-AD/eigen expressions? Would that do it?

I think the issue was the above and so it looked like the issue was in `matrix_set` but it was actually the weird logic bits

Reply:
@SteveBronder Awesome yes those looked fixed!

> I think the issue was the above and so it looked like the issue was in matrix_set but it was actually the weird logic bits

Looks like you made the `matrix_set` change, was that unintentional?

I noticed some odd behavior around ternary expressions, though it's not as critical because they only over-demote:
```
    // all of these demote everything including LHS and row_soa
    matrix[10, 10] ternary_soa = 1 ? row_soa : row_soa;
    matrix[10, 10] ternary_mixed_aos = 1 ? row_soa : user_func_aos;
    matrix[10, 10] ternary_mult_soa = row_soa * (1 ? row_soa : row_soa);
```

Reply:
> Looks like you made the matrix_set change, was that unintentional?

It was intentional, looking at it I think you are right that the way I was doing that could be wrong so I changed it like you said as I think that makes more sense.

Yes for ternary expressions C++ demands that the LHS and RHS types of the ternary operator match exactly so you can't have one side returning an SoA matrix with the other returning an AoS

Reply:
Gotcha. I see how it's hard to just demote the ternary subexpression with how it's set up. We could have it only demote when either side is AoS, but again it's not critical. In any case happy to call it.

Reply:
^So since they both have to match and can be confusing I forcefully demote both sides for all ternary expressions. I think at another time I want to come back to this and make these nicer

Reply:
Yes absolutely lets rock!!

Reply:
(fyi I just updated the PR to move the soa optim to only O1 so I think we should be good to go)